### PR TITLE
feat: implement handle_cancelled/3 worker callback

### PR DESCRIPTION
**Pull request by oliveigah**
closes https://github.com/dashbitco/nimble_pool/issues/43

### CODE REVIEW

File: lib/nimble_pool.ex [hunk 1]
```diff
@@ -745,19 +769,38 @@ defmodule NimblePool do
...
+          apply_worker_callback(worker, :handle_cancelled, args)
+        end
+
         state = remove_request(state, ref, mon_ref)
         {:noreply, remove_worker(reason, worker_server_state, state)}
 
       # The client timed out, sent us a message, and we dropped the deadlined request
       %{} ->
+        if function_exported?(worker, :handle_cancelled, 3) do
+          args = [nil, pool_state, :queued]
```
Comment (oliveigah) [inline lib/nimble_pool.ex:800]: Should we have other context argument for this scenario?

File: lib/nimble_pool.ex [hunk 2]
```diff
@@ -297,14 +297,38 @@ defmodule NimblePool do
...
+  happened the worker_state argument will be `nil`.
+
+  * `:checked_out` means the cancellation happened after resource checkout. This may happen
+  when the function given to `checkout!/4` raises.
+
+  This callback is optional.
+  """
+  @doc callback: :worker
+  @callback handle_cancelled(
+              worker_state :: worker_state | nil,
```
Comment (josevalim) [inline lib/nimble_pool.ex:318]: Let's have the context as first argument. I am wondering if we should pass the worker state. Maybe it is better to pass only `pool_state` and call it a pool callback.

Comment (oliveigah) [reply lib/nimble_pool.ex:318]: I think it may be useful for analyze some property of the worker state that may be causing the raises, but this can also be achieved be the terminate worker callback. 

Although on the cancelled callback might be more convenient since you know for sure that this callback will be called with `checked_out` context only when something really unexpected happens, therefore you do not need to handle special cases like on the terminate worker callback.

In summary I think it may be more convenient to keep it but I understand if you want to remove. I'm good to go whatever you pick, just let me know!

Comment (oliveigah) [reply lib/nimble_pool.ex]: Done! Let me know if you see any other improvement.

**Comment by josevalim** [comment]
Thank you! Please give it a try on Finch and, if all good, I can ship a new version.

**Comment by oliveigah** [comment]
@josevalim, as confirmed by @pcorey (https://github.com/sneako/finch/pull/267) this change have worked! 

Can you release a new version so we can get it into finch main? :pray:

**Comment by josevalim** [comment]
Done!
